### PR TITLE
1070 android tls

### LINKED ISSUE #1070: Android requires permissive cors settings -> security issue?
remote-server on Android runs on localhost and thus requires permissive cors settings, check if this is a security problem.

### PR DESCRIPTION
**Pull request by clemens-msupply**
- Add option to configure the certs directory
- Use rustls instead of openssl to avoid openssl issues on Android

Closes #1070

### CODE REVIEW

File: server/src/lib.rs [hunk 1]
```diff
@@ -341,32 +355,67 @@ pub async fn start_server(
...
 }
 
-const PRIVATE_CERT_FILE: &str = "./certs/key.pem";
-const PUBLIC_CERT_FILE: &str = "./certs/cert.pem";
-fn find_certs() -> ServerCertType {
-    if !Path::new(PRIVATE_CERT_FILE).exists() || !Path::new(PUBLIC_CERT_FILE).exists() {
+pub const PRIVATE_CERT_FILE: &str = "key.pem";
+pub const PUBLIC_CERT_FILE: &str = "cert.pem";
+
+fn find_certs(cert_dir: &str) -> ServerCertType {
```
Comment (andreievg) [inline server/src/lib.rs:374]: Would prefer a seperate file for cert specific methods.

Comment (clemens-msupply) [reply server/src/lib.rs]: [d468c5f](https://github.com/openmsupply/remote-server/pull/1086/commits/d468c5ffca6e8727516611e5680dc251f69267ca)

I moved all self signed cert related code. Think its much cleaner now, thanks!

File: android/src/android.rs [hunk 2]
```diff
@@ -56,6 +60,20 @@ pub mod android {
         })
     });
 
+    fn generate_certs(cert_dir: &PathBuf) {
+        let subject_alt_names = vec!["localhost".to_string()];
+        let cert = generate_simple_self_signed(subject_alt_names).unwrap();
```
Comment (clemens-msupply) [inline android/src/android.rs:65]: btw. this is the easiest way to generate a self signed cert I ever seen :-) By default it uses elliptic curve key pairs which are also very fast to generate.

Comment (jmbrunskill) [reply android/src/android.rs:65]: Yes, that's very nice!

File: server/Cargo.toml [hunk 3]
```diff
@@ -22,16 +22,18 @@ service = { path = "../service" }
 util = { path = "../util" }
 
 actix-cors = "0.6.1"
-actix-web = { version= "4.0.1", features = ["openssl"] } 
+actix-web = { version= "4.0.1", features = ["rustls"] } 
 actix-files = "0.6.0"
 openssl = { version = "0.10", features = ["v110"] }
```
Comment (andreievg) [inline server/Cargo.toml:27]: can we remove this now ?

**Review by andreievg — APPROVED** [review]
Looks good, thanks, just removal of openssl from deps would be great

File: server/src/self_signed_certs.rs [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+/// Load rustls server config
+pub fn load_self_signed_certs_rustls(
+    cert_files: SelfSignedCertFiles,
+) -> Result<ServerConfig, anyhow::Error> {
+    let certfile = std::fs::File::open(&cert_files.public_cert_file)?;
+    let mut reader = BufReader::new(certfile);
+    let certs = rustls_pemfile::certs(&mut reader)
+        .unwrap()
+        .iter()
```
Comment (andreievg) [inline server/src/self_signed_certs.rs:43]: this is probably stylistic thing, but I personally like into_iter/cloneless, becomes:

 .into_iter()
.map(rustls::Certificate)
.collect();

Comment (clemens-msupply) [reply server/src/self_signed_certs.rs]: thanks, think I actually copied it from the rustls examples :)

[7c5e2a7](https://github.com/openmsupply/remote-server/pull/1086/commits/7c5e2a7443a39e836e5758f31da92e533a8ebee9)

**Comment by clemens-msupply** [comment]
> Looks good, thanks, just removal of openssl from deps would be great

Sorry forgot to mention. I tried this but unfortunately it didn't make the Android build happy. Some dependency still uses openssl and if I don't bundle it explicitly it can't find the shared libs when doing the cross compilation... Bit annoying.

**Comment by andreievg** [comment]
> Sorry forgot to mention. I tried this but unfortunately it didn't make the Android build happy. Some dependency still uses openssl and if I don't bundle it explicitly it can't find the shared libs when doing the cross compilation... Bit annoying.

Is this only when compiling for android targets ? (i tried removing openssl and associated reference in android features), and managed to build from android folder (for default target)

**Comment by clemens-msupply** [comment]
> > Sorry forgot to mention. I tried this but unfortunately it didn't make the Android build happy. Some dependency still uses openssl and if I don't bundle it explicitly it can't find the shared libs when doing the cross compilation... Bit annoying.
> 
> Is this only when compiling for android targets ? (i tried removing openssl and associated reference in android features), and managed to build from android folder (for default target)

It only if you compile for a different target architecture...

**Pull request merged**